### PR TITLE
Add selectors for text and arbitrary objects

### PR DESCRIPTION
**Pull request by thomasloven**
## Proposed change

Add selectors for multi-line text and arbitrary objects.

Requires home-assistant/core#45112

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-selector/ha-selector-text.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  @property() public value?: any;
+
+  @property() public label?: string;
+
+  protected render() {
+    return html`<paper-textarea
+      .label=${this.label}
+      .value="${this.value}"
+      @keydown=${this._ignoreKeydown}
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:18]: Why is this needed?

**Review by bramkragten — APPROVED** [review]

File: src/components/ha-selector/ha-selector-text.ts (hunk 1 above)
Comment (thomasloven) [reply src/components/ha-selector/ha-selector-text.ts:18]: I don't know. It was in the markdown card editor, and I assumed it had to do with the quick select menu perhaps?

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-text.ts:18]: I think that is for the dialog, but this is not in a dialog, so we can leave it out.

Comment (thomasloven) [reply src/components/ha-selector/ha-selector-text.ts]: Could be. I removed it and don't run into any problems when I test.

But on the other hand, I also removed it from the markdown card editor, and I don't run into any problems there either...

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-text.ts]: press enter in the markdown editor ;-)

Comment (thomasloven) [reply src/components/ha-selector/ha-selector-text.ts]: Yes. That's what I though'd be the problem, and I expected it to close the editor... but it still didn't when I removed the stopPropagation.
Maybe it's a browser dependent thing? I only tried it on macOS Chrome and iOS Safari.

Either way, if that *is* the problem, it should stay in the markdown editor, and be left out here.

Comment (bramkragten) [reply src/components/ha-selector/ha-selector-text.ts]: And that is why we need comments 😄

File: src/components/ha-selector/ha-selector-text.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@customElement("ha-selector-text")
+export class HaTextSelector extends LitElement {
+  @property() public hass!: HomeAssistant;
+
+  @property() public value?: any;
+
+  @property() public label?: string;
+
+  protected render() {
+    return html`<paper-textarea
```
Comment (balloob) [inline src/components/ha-selector/ha-selector-text.ts:15]: I've updated the backend PR to make `text` selector have a `multiline` attribute. So we should only render a `textarea` when `multiline` is true.

Comment (thomasloven) [reply src/components/ha-selector/ha-selector-text.ts]: Fixed!

File: src/components/ha-selector/ha-selector-text.ts [hunk 3]
```diff
@@ -11,15 +13,25 @@ export class HaTextSelector extends LitElement {
...
+        autocapitalize="none"
+        autocomplete="off"
+        spellcheck="false"
+      ></paper-textarea>`;
+    }
+    return html`<paper-input
+      required
+      .value=${this.value}
+      @value-changed=${this._handleChange}
+      no-label-float
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:33]: `.label=${this.label}`

File: src/components/ha-selector/ha-selector-text.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:33]:
```suggestion
      .label=${this.label}
```

Comment (thomasloven) [reply src/components/ha-selector/ha-selector-text.ts:33]: Done

File: src/components/ha-selector/ha-selector-text.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (bramkragten) [inline src/components/ha-selector/ha-selector-text.ts:33]: This should be removed now?

**Review by bramkragten — APPROVED** [review]
